db/write_controller.cc

```diff
@@ -62,49 +61,39 @@ uint64_t WriteController::GetDelay(SystemClock* clock, uint64_t num_bytes) {
   // interval.
   auto time_now = NowMicrosMonotonic(clock);
 
-  uint64_t sleep_debt = 0;
-  uint64_t time_since_last_refill = 0;
-  if (last_refill_time_ != 0) {
-    if (last_refill_time_ > time_now) {
-      sleep_debt = last_refill_time_ - time_now;
-    } else {
-      time_since_last_refill = time_now - last_refill_time_;
-      bytes_left_ +=
-          static_cast<uint64_t>(static_cast<double>(time_since_last_refill) /
-                                kMicrosPerSecond * delayed_write_rate_);
-      if (time_since_last_refill >= kRefillInterval &&
-          bytes_left_ > num_bytes) {
-        // If refill interval already passed and we have enough bytes
-        // return without extra sleeping.
-        last_refill_time_ = time_now;
-        bytes_left_ -= num_bytes;
-        return 0;
-      }
+  const uint64_t kMicrosPerSecond = 1000000;
+  const uint64_t kMicrosPerRefill = 1024U;
+
+  if (next_refill_time_ == 0) {
+    // Start with an initial allotment of bytes for one interval
+    next_refill_time_ = time_now;
+  }
+  if (next_refill_time_ <= time_now) {
+    // Refill based on time interval plus any extra elapsed
+    uint64_t elapsed = time_now - next_refill_time_ + kMicrosPerRefill;
+    bytes_left_ += static_cast<uint64_t>(
+        1.0 * elapsed / kMicrosPerSecond * delayed_write_rate_ + 0.999999);
+    next_refill_time_ = time_now + kMicrosPerRefill;
+
+    if (bytes_left_ >= num_bytes) {
+      // Avoid delay if possible, to reduce DB mutex release & re-aquire.
+      bytes_left_ -= num_bytes;
+      return 0;
     }
   }
 
-  uint64_t single_refill_amount =
-      delayed_write_rate_ * kRefillInterval / kMicrosPerSecond;
-  if (bytes_left_ + single_refill_amount >= num_bytes) {
-    // Wait until a refill interval
-    // Never trigger expire for less than one refill interval to avoid to get
-    // time.
-    bytes_left_ = bytes_left_ + single_refill_amount - num_bytes;
-    last_refill_time_ = time_now + kRefillInterval;
-    return kRefillInterval + sleep_debt;
-  }
+  // We need to delay to avoid exceeding write rate.
+  assert(num_bytes > bytes_left_);
+  uint64_t bytes_over_budget = num_bytes - bytes_left_;
+  uint64_t needed_delay = static_cast<uint64_t>(
+      bytes_over_budget / static_cast<long double>(delayed_write_rate_) *
+      kMicrosPerSecond);
```

Comment:
Is the math simpler as "kMicrosPerSecond * bytes_over_budget / delated_write_rate_?  Would you still need the casts to double?

Reply:
Arithmetic overflow on 18TB write batch. Probably OK. 👍

Reply:
Actually, float division is faster than (large) scalar division. We don't need super accuracy float is fine.